### DISCUSSION TITLE
UI rendering is becoming extremely convoluted and complicated

### DISCUSSION THREAD

**Discussion by joekrill** [ory/kratos#4152] [Q&A] [answered]
Is there a plan to simplify how one should be rendering the various UIs? Or at least document it? 

I've noticed it has gotten _increasingly_ convoluted and complicated understanding when to render which nodes. And it's not documented anywhere, as far as I can tell. Unless you count the looking at the source code in the`@ory/elements` package - and even there it's not straight-forward at all:

Usually (but not always!) the `default` group should be included when filtering nodes. Same with "hidden" nodes. Sometimes `submit` and `button` nodes should be excluded.  `input` and `script` nodes should  always be included only when rendering the `webauthn` or `totp` groups. And there's just a _ton_ of this sorts of nuances, it seems. 

One of the main reasons to use Kratos is the "bring your own UI" aspect. But it's becoming kind of absurdly complicated to even do that. 

Is this documented somewhere that I just haven't found? Is there a plan to simplify all this without having to use `@ory/elements`?

**Comment by aeneasr** [comment, marked as answer]
hey @joekrill - thank you for the feedback and we agree that there are some implicit implications in the node UIs. What would be great would be to collect all the inconsistencies/discrepancies here or in an issue to figure out what to do (document, improve, fix, ...). As the complexity of the project is increasing day-by-day, it's also for us challenging to keep track of all the (implicit) api contracts we have.

**Comment by joekrill** [comment]
Thanks so much for the quick reply! 

> As the complexity of the project is increasing day-by-day, it's also for us challenging to keep track of all the (implicit) api contracts we have.

IMHO, there just shouldn't be _any_ implicit contracts to keep track of in the first place. That's really the whole problem. Any contracts should be explicit in the response. 

I think -- ideally -- the flow should include all necessary information needed to render the UI, without _any_ complicated rules needed on the rendering side of things, _at all_. I suspect that has always been the intention, and that things just unintentionally devolved a bit. Maybe the structure of the response needs to be reworked? 

I don't have a lot of in-depth knowledge of the project, so my observations/opinions may be misguided, but I'll share them anyway in the off chance they are useful. 

As far as I can tell (based on the [`@ory/elements` code](https://github.com/ory/elements/blob/main/src/react-components/ory/user-auth-card.tsx)) there's a few key reasons all this additional logic is needed in the first place:

1. Certain nodes should not be rendered based on some user state (i.e. ["the current flow is a two factor flow if the user is logged in and has any of the second factor methods enabled."](https://github.com/ory/elements/blob/main/src/react-components/ory/user-auth-card.tsx#L245)
2. The UI should be rendered as multiple separate forms.
3. There is a desire to customize the order/grouping of the form components.

For (1), I don't see why those nodes should even be returned in the first place. The backend has all this information already. It should know if the user is logged in. It should know if any second factor methods are enabled. It should certainly know the flow type it's returning. Those are some of the conditions currently being checked manually by `@ory/elements`, but the backend knows all those and so it should know whether or not to even return those nodes in the first place.

For (2), the response could be reworked to explicitly return those form groups instead of just relying on using the `group` property and implicit rules. 

Something like this would provide some backward compatibility and be more explicit:

```js
{
  ui: {
    nodes: [/*...nodes go here -- ensure each has a unique ID...*/],
    forms: [
      { 
         id: "oidc",
         nodes: [/* a list of node IDs to include. Or just inline the nodes here instead, even if it means duplicating them */], 
         messages: [],
      }, 
      { 
         id: "totp",
         nodes: [/*...*/],
         messages: [],
      },
    ]
  }
}
```

(3) will still have to be left to the rendering, but if the node groups are more explicit it becomes a purely cosmetic/ordering issue. It won't be necessary to break things up manually and figure out what other hidden fields need to be included, etc. 

Like I said, I don't have a lot of in-depth knowledge of the project, and I haven't spent a _ton_ of time thinking about all this, so I'm sure things are a lot more complicated than I've made them out to be. But I don't see how "bring your own UI" can be a realistic feature without removing all of those implicit rules.

**Comment by m4ttheweric** [comment]
Just wanted to add here that it's been an uphill battle for me as well trying to build out a custom UI. My wish here would be that there is no "default" group, and that each "group" could be mapped 1:1 to a `<form/>` element. And each group would contain every node it needs. 

Currently you just have to know that when rendering the "password" form/group that you have to go and snag the "identifier" node from the "default" group. The only way I figured that out was to look at the html page source for ory/elements and noticing that the password `<form />` included the identifier. 

I am trying to drink the koolaid as much as possible so the UI can be adaptive to whatever our server team builds, but it is really not easy and has required a lot of thought on developing flexible and adaptable approaches on the UI which has not been fun.

TL;DR: Please make the UI nodes more straightforward to consume and mappable 1:1 to a form.

**Reply by aeneasr** [reply]
Thank you for the feedback, I think Kratos should order the nodes with a default best practice way in the future. Anyone wanting other sorting can implement it as they wish. WDYT?
